src/mem.rs

```diff
@@ -1704,7 +1704,10 @@ pub fn convert_utf16_to_utf8(src: &[u16], dst: &mut [u8]) -> usize {
 /// not allocating memory for the worst case up front. Specifically,
 /// if the input starts with or ends with an unpaired surrogate, those are
 /// replaced with the REPLACEMENT CHARACTER.
-pub fn convert_utf16_to_str_partial(src: &[u16], dst: &mut str) -> (usize, usize) {
+/// Safety
+/// The caller must ensure that the content of the slice is valid UTF-8 before the borrow ends and the underlying str is used.
+/// Use of a str whose contents are not valid UTF-8 is undefined behavior.
```

Comment:
This safety contract doesn't make sense to be. The first sentence talks about "before the borrow ends," but presumably this is referencing something inside the implementation of the function that the caller doesn't control at all?

Moreover, your documentation here says that the caller shouldn't pass a `str` that isn't valid UTF-8, but that doesn't make any sense. A `str` is guaranteed to be valid UTF-8 and it is permissible for `unsafe` code to rely on that fact. The caller doesn't need to "do" anything for that. If the caller's str is invalid UTF-8, then it's either the caller's fault or some other piece of code. But it's certainly not required for this routine to be marked `unsafe` in order to assume that invariant.